### ISSUE TITLE
[New Concept Docs]: Context Manager Customization

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3081]
This issue describes how to implement the `context manager customization`  **concept docs**.
 The related **concept exercise** issue can be found [here](https://github.com/exercism/python/issues/3090).

 If you have not yet contributed to concept documents, this issue will require some upfront reading to give you the needed background knowledge.

## ✅  Getting started

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. 

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concepts:**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 

<br>

## 🎯  Goal

The goal of the documents described in this issue is to explain the concept of `context- manager customization` (_the machinery behind the with keyword_) in Python.

<br>

## 💡  Learning objectives

- Review what the `with` keyword *is* and what it _does_ in Python
  - context-handling & scope
  - setup and teardown processes
  - memory and resource allocation
- Understand different `context-manager` Types and their uses
- Get familiar with some of the  utilities & decorators in `contextlib`, and how they're used to customize a `context-manager`
  - `@contextlib.contextmanager`
  - `contextlib.closing()`
  - `contextlib.nullcontext()`
  - `contextlib.suppress`
  - `conextlib.redirect_stdout()`
  - `contextlib.redirect_stderr()`
  - `class contextlib.ContextDecorator`
- Understand how customized `context-managers` are used with the `with` statement
    - `__enter__`
    - `__exit__` 
- Implement some of the most common use cases for managing / customizing management of resources with the `with` keyword (_maybe some of the recipes listed in the [docs](https://docs.python.org/3/library/contextlib.html#examples-and-recipes)?_)
  - file opening/closing/processing
  - database connectivity
  - error handling
  - logging
  - testing

## 🚫  Topics that are Out of scope

While many of these topics could be considered prerequisites or necessary building blocks for this concept, we've marked them as "out of scope".  

This means that we'd like you to focus instruction and examples  primarily on the topic of **this** concept (_context managers and customization_), and save any specific instruction or custom use of the topics below for exercises that focus on those concepts in detail.  **For example:**  `@contextmanager` requires a `generator` to act as a factory.  But this exercise should not teach the making of or customization of a `generator`, since `generators` have their own exercise.  The same goes for creating `classes`,  `raising` & `handling` errors, or using `assert` for testing.  The overall goal should be to use _just enough_ to explain/instruct the student in the `context-manager` customization concept, and assume they have learned or will learn the other details.  If you have additional questions, **please reach out to one of the maintainers.**

<br>
<details>
<summary><b>Out of Scope Topics for Context Manager Customization</b></summary>

- `async` versions of `context-managers` (_we are leaving `async-io` and `async` off our concept list right now due to their complexity_)
- `classes`  & `class customization`
- `class-inheritance` 
- `comprehensions` 
- `comprehensions` in `lambdas`
- `coroutiens` 
- `decorators` 
- `error-handling` 
- `raising errors`
- `functions` and `higher-order functions`  
-  `functools` and related `map()`, `filter()` and `functools.reduce()`
- `generators`
- `lambdas` 
- using an `assignment expression` or "walrus" operator (`:=`) 
- `enums`
- `testing` and `assert`

</details>

<br>

## 🤔  Concepts

<details>
<summary><b>Concepts/Related Concepts</b></summary>

- `with`
- `context-managers`
- `contextlib`
-  Special Methods
   -   `__enter__`
   -  `__exit__`

</details>

<br>

## ↩️  Prerequisites

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  

<details>
<summary>Prereqs</summary>

- `basics`
- `bools`
- `classes`  
- `class customization` 
- `class-inheritance` 
- `comparisons`
- `rich-comparisons`
- `decorators`
- `dicts`
- `dict-methods`
- `raising-and-handling-errors` 
- `functions`
- `generators`
- `higher-order-functions`
- `lists`
- `list-methods`
- `loops`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `testing`
- `tuples`
- `with-statement`

</details>

<br>  

## 📚  Resources to refer to

<details>
<summary><b>Resources</b></summary>

- [Python Docs: Compound Statements](https://docs.python.org/3/reference/compound_stmts.html)
- [Python Docs:  With Statement Context Managers](https://docs.python.org/3/reference/datamodel.html#with-statement-context-managers)
- [Python Docs: With Statement](https://docs.python.org/3/reference/compound_stmts.html#the-with-statement)
  - [Python Docs: Reading and Writing Files](https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files)
- [Python Docs:  Context Manager Types](https://docs.python.org/3/library/stdtypes.html#context-manager-types)
- [PEP 0343: The `with` Statement](https://www.python.org/dev/peps/pep-0343/)
- [PEP 0343:  Examples](https://www.python.org/dev/peps/pep-0343/#examples)
- [Python Docs: `contextlib`](https://docs.python.org/3/library/contextlib.html#module-contextlib)
- [Python Module of the Week: `contextlib`](https://pymotw.com/3/contextlib/)
- [Python Morsels:  Context-managers](https://www.pythonmorsels.com/topics/context-managers/)
-  [Dan Bader: Context Managers and the `with` statement in Python](https://dbader.org/blog/python-context-managers-and-with-statement)
- [Real Python: Context managers](https://realpython.com/python-with-statement/#:~:text=Coding%20Class%2DBased%20Context%20Managers,-To%20implement%20the%20context%20management)
- [Martin Heinz: The Magic of Python Context Managers](https://martinheinz.dev/blog/34)
- [Innoplex: Fun With Context Managers](https://www.innoplexllc.com/post/python-fun-with-context-managers)
- [Siv Scripts:  A Gentle Introduction to Context Managers](https://alysivji.github.io/managing-resources-with-context-managers-pythonic.html)
- [arnavk.org: Context Managers in the Real World](http://arnavk.org/posts/python-context-managers/)
- [Abhinav Gupta:  Go Antipatterns - With* Context Managers](https://abhinavg.net/posts/context-manager-antipattern/)

</details>
<details>
<summary><b>Additional Articles</b></summary>

-  [David Beazley at (Unenix.org) Python: With That - 5 Easy Context Managers ](https://www.usenix.org/system/files/login/articles/12_beazley-online.pdf)
- [Jeff Knupp: Python with Context Managers](https://jeffknupp.com/blog/2016/03/07/python-with-context-managers/)
-  [Red's Digressions: The Curious Case of Python's Context Manager](https://rednafi.github.io/digressions/python/2020/03/26/python-contextmanager.html)
- [Cambridge Spark: Python Context manager](https://blog.cambridgespark.com/python-context-manager-3d53a4d6f017)

</details>

<br>

## 📁  Files to Be Created

Please see the following for more details on these files:  [concepts](https://exercism.org/docs/building/tracks/concepts)

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

* ### Concept `about.md`

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-aboutmd)

  * This can also be a summary/paraphrase of the  `about.md` document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the associated concept or practice exercises.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Concept `.meta/config.json` Entries
 
  For more information, see [Concept `.meta/config.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-metaconfigjson)

  * This file is likely already stubbed out.  Remember to add a concept **blurb** of less than 350 characters.  Please also add your GitHub username to the "authors" array, and any contributor GitHub usernames to the "contributors" array.

<br>

## 🎶  Implementation Notes

- Example code  should **only use syntax & concepts introduced within these docs or one of the prerequisite concept exercises or documents.** Where possible, please use REPL formatting, unless you are demonstrating pseudo code or a long code block.
  Please **do not use** syntax not previously covered in prerequisite topics or exercises.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

##  🆘  Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.
